### ISSUE TITLE
Support for Fine-tuning

### ISSUE THREAD

**Issue by quantumcoder121** [heyoeyo/muggled_sam#9]
Hi there!
So we are using this model and this repository in particular for our Research on a novel problem statement and would like to fine tune on medical datasets. Can you please provide support for finetuning? Also if already provided, can you guide how to finetune a pipeline created out of this model?

**Comment by heyoeyo** [comment]
Cool, that's awesome to hear!

For the time being, there's no fine-tuning code in this repo and probably won't be for at least a few months (if ever). Model documentation is the main focus atm, and that will take a while.

However, since this repo just loads SAMv1/SAMv2 weights, it will work with fine-tuned weights as well. The weights just need to be for the exact same model implementation (i.e. no added features/functionality).
So for example, if you're using SAMv2 weights, then it's possible to use the [official fine-tuning code](https://github.com/facebookresearch/sam2/tree/main/training#getting-started) to create a fine-tuned model and those weights will work with this repo as well. SAMv1 doesn't seem to have official fine-tuning code, but there are a number of unofficial implementations (like [this one](https://github.com/bhpfelix/segment-anything-finetuner)).

Alternatively, if you're interested in implementing the training yourself, a good starting point would be something similar to this [block of code](https://github.com/DepthAnything/Depth-Anything-V2/blob/31dc97708961675ce6b3a8d8ffa729170a4aa273/metric_depth/train.py#L121-L136) (from a very different model, but the idea is the same), except the 'criterion' would be replaced with the [SAM loss functions](https://github.com/facebookresearch/sam2/blob/main/training/loss_fns.py). Of course the official SAMv2 training code is also a good reference, and it does [something similar](https://github.com/facebookresearch/sam2/blob/c2ec8e14a185632b0a5d8b161928ceb50197eddc/training/trainer.py#L740-L792), but it's a lot harder to follow.

**Comment by heyoeyo** [comment]
Oh I see, there are probably a few different approaches...

Regarding the `encode_prompts` function, an alternative approach would be to use the `forward` function of the model which doesn't have the `torch.inference_mode()` wrapper and also runs the whole model for image segmentation. The only catch is that the prompts (e.g. points) need to be given as tensors (on the same device/dtype as the model) instead of lists and need a batch dimension added (see the [prepare_points](https://github.com/heyoeyo/muggled_sam/blob/fa5015d1a5f58646fcea5e21adf907872d0412a8/lib/v2_sam/coordinate_encoder_model.py#L139) function for example). So that's one way to get things working without any other changes.

To only train the coordinate encoder, you'd also need to 'freeze' all the other model weights before training. I think the simplest way to do this would be to freeze all weights and then unfreeze the coordinate encoder with something like:
```python
# Disable gradient tracking for all model weights
_, sam_model = make_sam_from_state_dict(...)
for p in sam_model.parameters():
  p.requires_grad = False

# Re-enable gradients for coordinate encoder only
for p in sam_model.coordinate_encoder.parameters():
  p.requires_grad = True
```
With this, running the model's `forward` function should be a bit like having everything except the coordinate encoder inside a `torch.inference_mode` block.

If you wanted to use the existing SAMv2 fine tuning code, you could do mostly the same thing, except freezing/unfreezing would use different names. You could do it with something like:
```python
# Freeze the whole model
predictor = SAM2ImagePredictor(...)
for p in predictor.model.parameters():
  p.requires_grad = False

# Unfreeze the weights used to encode coordinates
# (this has to be done directly, because the original weights not considered trainable by default)
predictor.model.sam_prompt_encoder.pe_layer.positional_encoding_gaussian_matrix.requires_grad = True
```

As far as surrogates are concerned, I think that can make a lot of sense too depending on how the code is setup (and if you're planning to do other forms of targeted training in the future). The simplest thing would be to just have a standalone function. For example, you could have a function which does the same prompt encoding as the original model, but with the coordinate encoder outside the inference mode block with something like:
```python
# Alternate version of 'encode_prompts' function which exists outside
# the sam model. With coord. encoder setup to track gradients for training
def trainable_encode_prompts(sam_model, box_list, fg_list, bg_list):

  boxes = sam_model.coordinate_encoder.prepare_boxes(box_list)
  fg, bg = sam_model.coordinate_encoder.prepare_points(fg_list, bg_list)
  boxenc, fgenc, bgenc = sam_model.coordinate_encoder(boxes, fg, bg)

  with torch.inference_mode():
    encoded_prompts = sam_model.prompt_encoder(boxenc, fgenc, bgenc)

  return encoded_prompts
```
Then instead of calling `sam_model.encode_prompts(...)` you could use `trainable_encode_prompts(sam_model, ...)`. Functions like these (which take the model as an input and run it differently) might make sense if you have other targeted/training-specific use cases in the future, since they're a bit more self contained/self-documenting.

**Comment by heyoeyo** [comment]
For anyone finding this in the future, it's now possible to directly disable the `inference_mode` blocks, which can save the trouble of having to make a separate function for training. This is done with something like:
```python
model.toggle_inference_mode(False)
```

There's also a [model_distillation](https://github.com/heyoeyo/muggled_sam/blob/main/simple_examples/model_distillation.py) example script that might be useful as a starting point for training/fine-tuning.
